packages/rule-https-only/src/rule.ts

```diff
@@ -0,0 +1,92 @@
+/**
+ * @fileoverview sonarwhal rule that checks if your site is using HTTPS and if it has mixed content.
+ */
+
+import { Category } from 'sonarwhal/dist/src/lib/enums/category';
+import { RuleContext } from 'sonarwhal/dist/src/lib/rule-context';
+
+import { IRule, RuleMetadata, FetchEnd, Response } from 'sonarwhal/dist/src/lib/types';
+import { debug as d } from 'sonarwhal/dist/src/lib/utils/debug';
+import { RuleScope } from 'sonarwhal/dist/src/lib/enums/rulescope';
+import { isHTTPS } from 'sonarwhal/dist/src/lib/utils/misc';
+
+const debug: debug.IDebugger = d(__filename);
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+export default class HttpsOnlyRule implements IRule {
+    private targetIsServedOverHTTPS: boolean = false;
+
+    public static readonly meta: RuleMetadata = {
+        docs: {
+            category: Category.security,
+            description: `sonarwhal rule that checks if your site is using HTTPS and if it has mixed content.`
+        },
+        id: 'https-only',
+        schema: [],
+        scope: RuleScope.site
+    }
+
+    public constructor(context: RuleContext) {
+        let target: string;
+
+        const validateTarget = async (fetchEvent: FetchEnd): Promise<void> => {
+            const { resource } = fetchEvent;
+
+            if (!isHTTPS(resource)) {
+                debug('HTTPS no detected');
+
+                await context.report(resource, null, 'The site should be HTTPS');
+
+                return;
+            }
+
+            debug('HTTPS detected');
+
+            this.targetIsServedOverHTTPS = true;
+
+            return;
+        };
+
+        const anyHopHTTP = (response: Response) => {
+            const { hops } = response;
+
+            return hops.some((hop) => {
+                return !isHTTPS(hop);
+            });
+        };
+
+        const validate = async (fetchEnd: FetchEnd) => {
+            // We are assuming the first `fetch::end` event recieved is the target url.
+            if (!target) {
+                target = fetchEnd.resource;
+
+                await validateTarget(fetchEnd);
+
+                return;
+            }
+
+            if (!this.targetIsServedOverHTTPS) {
+                return;
+            }
+
+            const { resource, response } = fetchEnd;
+
+            if (!isHTTPS(resource)) {
+                await context.report(resource, null, 'Should be served over HTTPS');
+
+                return;
+            }
+
+            if (anyHopHTTP(response)) {
+                await context.report(resource, null, `Shouldn't be redirected from HTTP`);
+            }
+        };
+
+        context.on('fetch::end::*', validate);
```

Comment:
I don't think we are fetching `<audio>`, `<video>`, `<object>`..., so we are not validating a lot of things.

See: https://w3c.github.io/webappsec-mixed-content/

Reply:
I guess it will depend on the connector and if the case of media if it has `autoplay` enabled?

What I'm thinking on doing:
* Have a list with all the resources we've analyzed
* Subscribe to those elements events and check it there
* If we receive them via a `fetch::end`, use that information instead